monai/networks/blocks/selfattention.py

```diff
@@ -130,23 +145,28 @@ def forward(self, x):
             q = q.to(self.attention_dtype)
             k = k.to(self.attention_dtype)
 
-        att_mat = torch.einsum("blxd,blyd->blxy", q, k) * self.scale
+        if self.use_flash_attention:
+            x = F.scaled_dot_product_attention(q, k, v)
```

Comment:
We may also inlcude these args here to be consistent with before.
https://github.com/Project-MONAI/GenerativeModels/blob/7428fce193771e9564f29b91d29e523dd1b6b4cd/generative/networks/blocks/selfattention.py#L121-L123

Reply:
Scale and dropout are fine. I am a bit confused when it comes to causal and attn_bias. attn_bias is a mask, which is set to xops.LowerTriangularMask()  in the old implementation. Nonetheless, here you cannot set it up if causal is True. It's automatically derived. It seems that if causal is True, we can just pass "is_causal = True" and the function will derive the mask. However, when causal is not True, we can pass a mask, but in the init method, self.causal_mask is only set up when causal is True. I think to be safe, I'd just pass the causal value and not any mask.

Reply:
If you use causal = True, and you run with and without flash_attention, you do not get a similar output

Reply:
Also, the test test_causal cannot be reproduced for flash_attention because save_attn has to be False. There's no way the attention matrix can be derived and therefore, we cannot verify that it's masking things correctly.

Reply:
I also had to add self.dropout_rate as an attribute, as it's necessary for the call to the function.

Reply:
For the casual one, we can use `is_casual` in the torch: https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html

```
import torch
import xformers.ops as xops
import torch.nn.functional as F
from copy import deepcopy

torch.manual_seed(0)

query = torch.rand(8, 2560, 1, 32).to("cuda")
key = torch.rand(8, 2560, 1, 32).to("cuda")
value = torch.rand(8, 2560, 1, 32).to("cuda")

x = xops.memory_efficient_attention(query, key, value, attn_bias=xops.LowerTriangularMask())
print(x.squeeze()[0,0,...])
y = F.scaled_dot_product_attention(query.transpose(1, 2), key.transpose(1, 2), value.transpose(1, 2), is_causal=True).transpose(1, 2)
print(y.squeeze()[0, 0, ...])

torch.sum(torch.abs(x.squeeze()-y.squeeze()))
```